### ISSUE TITLE
Generated config sometimes unusable

### ISSUE THREAD

**Issue by jvstein** [rbtr/pachinko#6]
The config generated by `pachinko config` is not always usable. This is because the order of the `processors` list is non-deterministic and potentially reordered each time it is written.

The particular issue I ran into is that the `video-id` processor must be executed before the `tv-id` processor (because of the check in [video.go#L42](https://github.com/rbtr/pachinko/blob/master/plugin/processor/vidid/video.go#L42). If the processors are out of order, it will not work.

Example bad config created by the command:

```
dry-run: false
inputs:
- name: path
  src-dir: ./src
log-format: text
log-level: info
outputs:
- create-dirs: true
  dry-run: true
  name: move
  overwrite: false
- name: stdout
pipeline:
  buffer: 0
processors:
- api-key: ""
  name: tvdb
  request-limit: 10
- dest-dir: dest
  episode-names: false
  format: not-implemented
  name: tv-dest
  season-dirs: true
  tv-prefix: tv
- matchers:
  - (?i)(?:s+\d+)(?:\.|\s|-|_|x)*(e+\d+)
  - (?i)(?:season)+(?:(?:\.|\s|-|_|x|\/)\d+)
  - (?i)(?:[^[:alpha:]\d])\d{1,3}[x-]\d{1,3}
  name: tv-id
- matchers:
  - (?i)\b([\s\w.-]*)[\s.\/-]+(?:\((\d+)\))?[\s.\/-]?(\d{1,3})[x-](\d{1,3})
  - (?i)\b([\s\w.-]*?)?(?:[\s\(.\/-](\d{4})[\s\).\/-])?[\s\w.-]?(?:s+(\d+))(?:\.|\s|-|_|x)*(?:e+(\d+))
  - (?i)\b([\s\w.-]*)[\s.\/-]+(?:\((\d+)\))?[\s.\/-]?(?:season|series).?(\d+).?(?:episode)?[^\d(]?(\d+)
  name: tv-meta
  sanitize-name: true
- file-types:
  - .avi
  - .divx
  - .m4v
  - .mkv
  - .mov
  - .mp4
  - .xvid
  name: video-id
```

**Comment by rbtr** [comment]
Thanks for the report! 

`mapstructure` alphabetizes maps when they are serialized, so the plugins are always going to be in the same (wrong) order when the config is generated. I tried to emphasize that the order matters in the README, but this is obviously unintuitive behaviour that I would like to correct.

I have been noodling on some ideas to improve the data flow already. The original design intent was that the plugins in each stage would not depend on each other and could be executed in parallel, but it didn't quite come out that way with the bulk `processor` plugin group...

I'm working on a minor redesign which has pre- and post- processor designations to group plugins which will control the run order for dependencies like this. i.e. the `video-id` plugin would be a pre-processor, and the `tv-id` a (intra-)processor so that the former is guaranteed to execute before the latter. The config file would have subsections under `processor` for pre/post, and the plugins in the stage would still be expected to be parallelizable.

I pushed a draft of this to the `movies` branch. It (probably) doesn't run yet, but it looks like you have some Go background and might like to take a look. I also see that you found that `send on closed channel` panic when the buffer is > 1 also, please feel free to PR that! :grin: 

TL;DR: I'm working on a redesign to `processor` plugin layouts that will resolve this incidentally (-> `movies` branch). Thanks again for the report and feel free to contrib :+1:

**Comment by rbtr** [comment]
@jvstein #8 which was just pushed, and the v0.2.0 tag, should address the issue you've raised here. It would be great if you could take that for a spin and provide any feedback concerning this or any other issue you might encounter. Thanks!
